File: src/com/lushprojects/circuitjs1/client/ImageExporter.java
package com.lushprojects.circuitjs1.client;

import java.util.Arrays;

import com.google.gwt.canvas.client.Canvas;
import com.google.gwt.canvas.dom.client.Context2d;
import com.google.gwt.dom.client.CanvasElement;
import com.google.gwt.core.client.Callback;
import com.google.gwt.core.client.ScriptInjector;
import com.google.gwt.user.client.Window;

public class ImageExporter {

	static final int CAC_PRINT = 0;
	static final int CAC_IMAGE = 1;
	static final int CAC_SVG   = 2;

	CirSim sim;
	boolean loadedCanvas2SVG = false;

	ImageExporter(CirSim sim) {
		this.sim = sim;
	}

	void doExportAsImage() {
		sim.dialogShowing = new ExportAsImageDialog(CAC_IMAGE);
		sim.dialogShowing.show();
	}

	private static native void clipboardWriteImage(CanvasElement cv) /*-{
		cv.toBlob(function(blob) {
		    var promise = parent.navigator.clipboard.write([new ClipboardItem({ "image/png": blob })]);
		    promise.then(function(x) { console.log(x); });
		});
	}-*/;

	void doImageToClipboard() {
		Canvas cv = getCircuitAsCanvas(CAC_IMAGE);
		clipboardWriteImage(cv.getCanvasElement());
	}

	native void printCanvas(CanvasElement cv) /*-{
	    var img    = cv.toDataURL("image/png");
	    var win = window.open("", "print", "height=500,width=500,status=yes,location=no");
	    win.document.title = "Print Circuit";
	    win.document.open();
	    win.document.write('<img src="'+img+'"/>');
	    win.document.close();
	    setTimeout(function(){win.print();},1000);
	}-*/;

	void doPrint() {
	    Canvas cv = getCircuitAsCanvas(CAC_PRINT);
	    printCanvas(cv.getCanvasElement());
	}

	boolean initializeSVGScriptIfNecessary(final String followupAction) {
		// load canvas2svg if we haven't already
		if (!loadedCanvas2SVG) {
			ScriptInjector.fromUrl("canvas2svg.js").setCallback(new Callback<Void,Exception>() {
				public void onFailure(Exception reason) {
					Window.alert("Can't load canvas2svg.js.");
				}
				public void onSuccess(Void result) {
					loadedCanvas2SVG = true;
					if (followupAction.equals("doExportAsSVG")) {
						doExportAsSVG();
					} else if (followupAction.equals("doExportAsSVGFromAPI")) {
						doExportAsSVGFromAPI();
					}
				}
			}).inject();
			return false;
		}
		return true;
	}

	void doExportAsSVG() {
		if (!initializeSVGScriptIfNecessary("doExportAsSVG")) {
			return;
		}
		sim.dialogShowing = new ExportAsImageDialog(CAC_SVG);
		sim.dialogShowing.show();
	}

	public void doExportAsSVGFromAPI() {
		if (!initializeSVGScriptIfNecessary("doExportAsSVGFromAPI")) {
			return;
		}
		String svg = getCircuitAsSVG();
		sim.jsInterface.callSVGRenderedHook(svg);
	}

	public Canvas getCircuitAsCanvas(int type) {
	    	// create canvas to draw circuit into
	    	Canvas cv = Canvas.createIfSupported();
	    	Rectangle bounds = sim.getCircuitBounds();

		// add some space on edges because bounds calculation is not perfect
	    	int wmargin = 140;
	    	int hmargin = 100;
	    	int w = (bounds.width*2+wmargin) ;
	    	int h = (bounds.height*2+hmargin) ;
	    	cv.setCoordinateSpaceWidth(w);
	    	cv.setCoordinateSpaceHeight(h);

		Context2d context = cv.getContext2d();
		drawCircuitInContext(context, type, bounds, w, h);
		return cv;
	}

	// create SVG context using canvas2svg
	native static Context2d createSVGContext(int w, int h) /*-{
	    return new C2S(w, h);
	}-*/;

	native static String getSerializedSVG(Context2d context) /*-{
	    return context.getSerializedSvg();
	}-*/;

	public String getCircuitAsSVG() {
	    Rectangle bounds = sim.getCircuitBounds();

	    // add some space on edges because bounds calculation is not perfect
	    int wmargin = 140;
	    int hmargin = 100;
	    int w = (bounds.width+wmargin) ;
	    int h = (bounds.height+hmargin) ;
	    Context2d context = createSVGContext(w, h);
	    drawCircuitInContext(context, CAC_SVG, bounds, w, h);
	    return getSerializedSVG(context);
	}

	void drawCircuitInContext(Context2d context, int type, Rectangle bounds, int w, int h) {
		Graphics g = new Graphics(context);
		context.setTransform(1, 0, 0, 1, 0, 0);
	    	double oldTransform[] = Arrays.copyOf(sim.transform, 6);

	        double scale = 1;

		// turn on white background, turn off current display
		boolean p = sim.menus.printableCheckItem.getState();
		boolean c = sim.menus.dotsCheckItem.getState();
		boolean print = (type == CAC_PRINT);
		if (print)
		    sim.menus.printableCheckItem.setState(true);
	        if (sim.menus.printableCheckItem.getState()) {
	            CircuitElm.whiteColor = Color.black;
	            CircuitElm.lightGrayColor = Color.black;
	            g.setColor(Color.white);
	        } else {
	            CircuitElm.whiteColor = Color.white;
	            CircuitElm.lightGrayColor = Color.lightGray;
	            g.setColor(Color.black);
	        }
	        g.fillRect(0, 0, w, h);
		sim.menus.dotsCheckItem.setState(false);

	    	int wmargin = 140;
	    	int hmargin = 100;
	        if (bounds != null)
	            scale = Math.min(w /(double)(bounds.width+wmargin),
	                             h/(double)(bounds.height+hmargin));

	        // ScopeElms need the transform array to be updated
		sim.transform[0] = sim.transform[3] = scale;
		sim.transform[4] = -(bounds.x-wmargin/2);
		sim.transform[5] = -(bounds.y-hmargin/2);
		context.scale(scale, scale);
		context.translate(sim.transform[4], sim.transform[5]);
		context.setLineCap(Context2d.LineCap.ROUND);

		// draw elements
		for (CircuitElm ce : sim.elmList) {
		    ce.draw(g);
		}
		int i;
		for (i = 0; i != sim.postDrawList.size(); i++) {
		    CircuitElm.drawPost(g, sim.postDrawList.get(i));
		}

		// restore everything
		sim.menus.printableCheckItem.setState(p);
		sim.menus.dotsCheckItem.setState(c);
		sim.transform = oldTransform;
	}
}

File: ImageExporter.ts
/*
    Copyright (C) Paul Falstad and Iain Sharp

    This file is part of CircuitJS1.

    CircuitJS1 is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 2 of the License, or
    (at your option) any later version.

    CircuitJS1 is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CircuitJS1.  If not, see <http://www.gnu.org/licenses/>.
*/

import { CirSim } from "./CirSim";
import { Graphics } from "./Graphics";
import { CircuitElm } from "./CircuitElm";
import { Color } from "./Color";
import { Rectangle } from "./Rectangle";

export class ImageExporter {
    static readonly CAC_PRINT = 0;
    static readonly CAC_IMAGE = 1;
    static readonly CAC_SVG   = 2;

    sim: CirSim;

    constructor(sim: CirSim) {
        this.sim = sim;
    }

    static defaultFileName(ext: string): string {
        const d = new Date();
        const pad = (n: number) => (n < 10 ? "0" + n : "" + n);
        const dtf = "" + d.getFullYear() + pad(d.getMonth() + 1) + pad(d.getDate()) +
            "-" + pad(d.getHours()) + pad(d.getMinutes());
        return "circuit-" + dtf + ext;
    }

    doExportAsImage(): void {
        if (CirSim.isElectron()) {
            const dataURL = this.getCircuitAsCanvas(ImageExporter.CAC_IMAGE).toDataURL();
            const base64 = dataURL.substring(dataURL.indexOf(',') + 1);
            (window as any).showSaveDialog?.(ImageExporter.defaultFileName(".png")).then((file: any) => {
                if (file.canceled)
                    return;
                (window as any).saveFile(file, base64, "base64");
            });
            return;
        }
        CirSim.dialogShowing = new (window as any).ExportAsImageDialog(ImageExporter.CAC_IMAGE, this.sim);
        CirSim.dialogShowing.show();
    }

    doImageToClipboard(): void {
        const cv = this.getCircuitAsCanvas(ImageExporter.CAC_IMAGE);
        cv.toBlob((blob) => {
            if (blob == null)
                return;
            const item = new (window as any).ClipboardItem({ "image/png": blob });
            navigator.clipboard.write([item]).then((x: any) => console.log(x));
        });
    }

    doPrint(): void {
        const cv = this.getCircuitAsCanvas(ImageExporter.CAC_PRINT);
        const img = cv.toDataURL("image/png");
        const style = document.createElement("style");
        style.id = "circuit-print-style";
        style.innerHTML = "@media print { body > *:not(#circuit-print-overlay) { display: none !important; } } " +
            "#circuit-print-overlay { display: none; } " +
            "@media print { @page { size: auto; margin: 10mm; } " +
            "#circuit-print-overlay { display: block !important; width: 100%; height: 100%; } " +
            "#circuit-print-overlay img { max-width: 100%; max-height: 100%; width: auto; height: auto; " +
            "display: block; margin: 0 auto; page-break-inside: avoid; } }";
        document.head.appendChild(style);
        const overlay = document.createElement("div");
        overlay.id = "circuit-print-overlay";
        const imgEl = document.createElement("img");
        imgEl.src = img;
        overlay.appendChild(imgEl);
        document.body.appendChild(overlay);
        setTimeout(() => {
            window.print();
            document.body.removeChild(overlay);
            document.head.removeChild(style);
        }, 500);
    }

    doExportAsSVG(): void {
        if (CirSim.isElectron()) {
            (window as any).showSaveDialog?.(ImageExporter.defaultFileName(".svg")).then((file: any) => {
                if (file.canceled)
                    return;
                (window as any).saveFile(file, this.getCircuitAsSVG(), "utf8");
            });
            return;
        }
        CirSim.dialogShowing = new (window as any).ExportAsImageDialog(ImageExporter.CAC_SVG, this.sim);
        CirSim.dialogShowing.show();
    }

    doExportAsSVGFromAPI(): void {
        const svg = this.getCircuitAsSVG();
        this.sim.jsInterface.callSVGRenderedHook(svg);
    }

    getCircuitAsCanvas(type: number): HTMLCanvasElement {
        const cv = document.createElement("canvas");
        const bounds = this.sim.getCircuitBounds();

        // add some space on edges because bounds calculation is not perfect
        const wmargin = 140;
        const hmargin = 100;
        const w = bounds.width * 2 + wmargin;
        const h = bounds.height * 2 + hmargin;
        cv.width = w;
        cv.height = h;

        const context = cv.getContext("2d") as CanvasRenderingContext2D;
        this.drawCircuitInContext(context, type, bounds, w, h);
        return cv;
    }

    getCircuitAsSVG(): string {
        const bounds = this.sim.getCircuitBounds();

        // add some space on edges because bounds calculation is not perfect
        const wmargin = 140;
        const hmargin = 100;
        const w = bounds.width + wmargin;
        const h = bounds.height + hmargin;
        const context = new (window as any).C2S(w, h);
        this.drawCircuitInContext(context, ImageExporter.CAC_SVG, bounds, w, h);
        return context.getSerializedSvg();
    }

    drawCircuitInContext(context: CanvasRenderingContext2D, type: number, bounds: Rectangle, w: number, h: number): void {
        const g = new Graphics(context);
        context.setTransform(1, 0, 0, 1, 0, 0);
        const oldTransform = this.sim.transform.slice();

        let scale = 1;

        // turn on white background, turn off current display
        const p = this.sim.menus.printableCheckItem.getState();
        const c = this.sim.menus.dotsCheckItem.getState();
        const print = (type === ImageExporter.CAC_PRINT);
        if (print)
            this.sim.menus.printableCheckItem.setState(true);
        if (this.sim.menus.printableCheckItem.getState()) {
            CircuitElm.whiteColor = Color.black;
            CircuitElm.lightGrayColor = Color.black;
            g.setColor(Color.white);
        } else {
            CircuitElm.whiteColor = Color.white;
            CircuitElm.lightGrayColor = Color.lightGray;
            g.setColor(Color.black);
        }
        g.fillRect(0, 0, w, h);
        this.sim.menus.dotsCheckItem.setState(false);

        const wmargin = 140;
        const hmargin = 100;
        if (bounds != null)
            scale = Math.min(w / (bounds.width + wmargin), h / (bounds.height + hmargin));

        // ScopeElms need the transform array to be updated
        this.sim.transform[0] = this.sim.transform[3] = scale;
        this.sim.transform[4] = -(bounds.x - wmargin / 2);
        this.sim.transform[5] = -(bounds.y - hmargin / 2);
        context.scale(scale, scale);
        context.translate(this.sim.transform[4], this.sim.transform[5]);
        context.lineCap = "round";

        // draw elements
        for (const ce of this.sim.elmList)
            ce.draw(g);
        for (let i = 0; i !== this.sim.postDrawList.length; i++)
            CircuitElm.drawPost(g, this.sim.postDrawList[i]);

        // restore everything
        this.sim.menus.printableCheckItem.setState(p);
        this.sim.menus.dotsCheckItem.setState(c);
        this.sim.transform = oldTransform;
    }
}
